### PR TITLE
[Test] Restores the offsets transferred by MirrorMaker2 to the consumer group on the target cluster

### PR DESCRIPTION
**Pull request by michalxo**
### Type of change

New test `MirrorMaker2ST#testRestoreOffsetsInConsumerGroup` for Active/Active MM2 deployment, which simulates (not only) consuming part of messages from source cluster and continuing to receive mirrored messages by "noting" already consumed offsets from source cluster and reading only "remaining" offsets from target cluster. 
So there are no duplicated consumed messages when source cluster crashes and receiving continues from target cluster where left of.

Related https://cwiki.apache.org/confluence/display/KAFKA/KIP-545

### Checklist

- [x] Make sure all tests pass
- [x] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [x] Reference relevant issue(s) and close them after merging

### CODE REVIEW

File: systemtest/src/test/java/io/strimzi/systemtest/mirrormaker/MirrorMaker2ST.java [hunk 1]
```diff
@@ -914,4 +1086,13 @@ void removeResourcesAndTopics() {
...
     }
+
+    protected static void waitForReadinessOfKafkaCluster(String kafkaClusterName, int zookeeperPods, int kafkaPods, int entityOperatorPods) {
+        LOGGER.info("Waiting for Zookeeper StatefulSet");
+        StatefulSetUtils.waitForAllStatefulSetPodsReady(KafkaResources.zookeeperStatefulSetName(kafkaClusterName), zookeeperPods);
+        LOGGER.info("Waiting for Kafka StatefulSet");
+        StatefulSetUtils.waitForAllStatefulSetPodsReady(KafkaResources.kafkaStatefulSetName(kafkaClusterName), kafkaPods);
+        LOGGER.info("Waiting for EO Deployment");
+        DeploymentUtils.waitForDeploymentAndPodsReady(KafkaResources.entityOperatorDeploymentName(kafkaClusterName), entityOperatorPods);
+    }
```
Comment (im-konge) [inline systemtest/src/test/java/io/strimzi/systemtest/mirrormaker/MirrorMaker2ST.java:1097]: This is not needed -> you can use `KafkaUtils.waitForKafkaReady(clusterName);`

File: systemtest/src/test/java/io/strimzi/systemtest/mirrormaker/MirrorMaker2ST.java [hunk 2]
```diff
@@ -893,6 +901,170 @@ void testConfigureDeploymentStrategy() {
...
+        final String topicSourceNameMirrored = "test-sync-offset-" + new Random().nextInt(Integer.MAX_VALUE);
+        final String topicTargetNameMirrored = kafkaClusterSourceName + "." + topicSourceNameMirrored;
+        final String consumerGroup = "mm2-test-consumer-group";
+        final String sourceProducerName = "mm2-producer-source-" + ClientUtils.generateRandomConsumerGroup();
+        final String sourceConsumerName = "mm2-consumer-source-" + ClientUtils.generateRandomConsumerGroup();
+        final String targetProducerName = "mm2-producer-target-" + ClientUtils.generateRandomConsumerGroup();
+        final String targetConsumerName = "mm2-consumer-target-" + ClientUtils.generateRandomConsumerGroup();
+
+        // Deploy source kafka
+        KafkaResource.kafkaClient().inNamespace(ResourceManager.kubeClient().getNamespace()).createOrReplace(
```
Comment (im-konge) [inline systemtest/src/test/java/io/strimzi/systemtest/mirrormaker/MirrorMaker2ST.java:916]: You can do it by `KafkaResource.kafkaWithoutWait(...)` -> it's the same thing and we are using this in all test cases

File: systemtest/src/test/java/io/strimzi/systemtest/mirrormaker/MirrorMaker2ST.java [hunk 3]
```diff
@@ -893,6 +901,170 @@ void testConfigureDeploymentStrategy() {
...
+                KafkaResource.kafkaPersistent(kafkaClusterSourceName, 3, 3).build()
+        );
+        // Deploy target kafka
+        KafkaResource.kafkaClient().inNamespace(ResourceManager.kubeClient().getNamespace()).createOrReplace(
+                KafkaResource.kafkaPersistent(kafkaClusterTargetName, 3, 3).build()
+        );
+        // wait for Kafka clusters readiness
+        waitForReadinessOfKafkaCluster(kafkaClusterSourceName, 3, 3, 1);
+        waitForReadinessOfKafkaCluster(kafkaClusterTargetName, 3, 3, 1);
```
Comment (im-konge) [inline systemtest/src/test/java/io/strimzi/systemtest/mirrormaker/MirrorMaker2ST.java:925]: How I mentioned, use `KafkaUtils.waitForKafkaReady()`

File: systemtest/src/test/java/io/strimzi/systemtest/mirrormaker/MirrorMaker2ST.java [hunk 4]
```diff
@@ -893,6 +901,170 @@ void testConfigureDeploymentStrategy() {
...
+                    .endHeartbeatConnector()
+                    .withTopicsPattern(".*")
+                    .withGroupsPattern(".*")
+                .endMirror()
+                .endSpec().build());
+
+        // Deploy Topic
+//        KafkaTopicResource.createAndWaitForReadiness(KafkaTopicResource.topic(kafkaClusterSourceName, topicSourceNameMirrored, 3).build());
+
+        KafkaClientsResource.createAndWaitForReadiness(KafkaClientsResource.deployKafkaClients(false, clusterName + "-" + Constants.KAFKA_CLIENTS).build());
```
Comment (im-konge) [inline systemtest/src/test/java/io/strimzi/systemtest/mirrormaker/MirrorMaker2ST.java:983]: You don't need this when you're using example clients

File: systemtest/src/test/java/io/strimzi/systemtest/mirrormaker/MirrorMaker2ST.java [hunk 5]
```diff
@@ -893,6 +901,170 @@ void testConfigureDeploymentStrategy() {
...
+        ClientUtils.waitForClientSuccess(targetConsumerName, NAMESPACE, 40);
+        JobUtils.deleteJobWithWait(NAMESPACE, targetConsumerName);
+
+//        assertThat("Expected to receive 40 messages.", receivedMessagesTargetPart2 == 40);
+//        internalKafkaClientTarget.checkProducedAndConsumedMessages(sentMessagesSource, receivedMessagesTarget + receivedMessagesTargetPart2);
+    }
+
+    private String createNewConsumerCliendId(String prefix) {
+        return  "client.id=" + prefix + "-id-" + consumerCounter++;
+    }
```
Comment (im-konge) [inline systemtest/src/test/java/io/strimzi/systemtest/mirrormaker/MirrorMaker2ST.java:1058]: This is not needed FMPOV - you use it only in commented code, also you can use `ClientUtils.generateRandomConsumerGroup()` and pass it to `withConsumerGroup` or something.

Comment (michalxo) [reply systemtest/src/test/java/io/strimzi/systemtest/mirrormaker/MirrorMaker2ST.java:1058]: Yup, works without this anyway.

**Comment by michalxo** [comment]
@strimzi-ci run tests test_only profile=regression testcase=MirrorMaker2ST#testRestoreOffsetsInConsumerGroup

**Comment by michalxo** [comment]
@strimzi-ci run tests test_only profile=regression testcase=MirrorMaker2ST#testRestoreOffsetsInConsumerGroup

**Comment by im-konge** [comment]
@michalxo Can you please change PR name and add some description?

File: systemtest/src/test/java/io/strimzi/systemtest/mirrormaker/MirrorMaker2ST.java [hunk 6]
```diff
@@ -893,6 +900,161 @@ void testConfigureDeploymentStrategy() {
...
+        LOGGER.info("Receive 10 msgs from source cluster");
+        internalClientSourceJob = internalClientSourceJob.toBuilder().withMessageCount(10).withAdditionalConfig("max.poll.records=10").build();
+        internalClientSourceJob.createAndWaitForReadiness(internalClientSourceJob.consumerStrimzi().build());
+        ClientUtils.waitForClientSuccess(sourceConsumerName, NAMESPACE, 10);
+        JobUtils.deleteJobWithWait(NAMESPACE, sourceConsumerName);
+
+        waitForMM2OffsetIntervalMirroring(syncGroupOffsetsIntervalSeconds);
+
+        LOGGER.info("Receive 40 msgs from target cluster");
+        KafkaBasicExampleClients internalClientTargetJob = initialInternalClientTargetJob.toBuilder().withMessageCount(40).withAdditionalConfig("max.poll.records=40").build();
```
[the remark this answers is not in the document]
Comment (michalxo) [reply systemtest/src/test/java/io/strimzi/systemtest/mirrormaker/MirrorMaker2ST.java:1039]: Yeah, you're right. This does not need to be used.

File: systemtest/src/test/java/io/strimzi/systemtest/mirrormaker/MirrorMaker2ST.java [hunk 7]
```diff
@@ -893,6 +900,161 @@ void testConfigureDeploymentStrategy() {
...
+        internalClientTargetJob.createAndWaitForReadiness(internalClientTargetJob.consumerStrimzi().build());
+        ClientUtils.waitForClientSuccess(targetConsumerName, NAMESPACE, 40);
+        JobUtils.deleteJobWithWait(NAMESPACE, targetConsumerName);
+
+        // There should be no more messages to read
+        initialInternalClientTargetJob.createAndWaitForReadiness(initialInternalClientTargetJob.consumerStrimzi().build());
+        assertThrows(WaitException.class, () -> ClientUtils.waitForClientSuccess(targetConsumerName, NAMESPACE, 1));
```
[the remark this answers is not in the document]
Comment (michalxo) [reply systemtest/src/test/java/io/strimzi/systemtest/mirrormaker/MirrorMaker2ST.java]: I have added more logging information before each logical step. Please let me know, if it makes more sense now.
Thank you.

**Comment by michalxo** [comment]
@strimzi-ci run tests test_only profile=regression testcase=MirrorMaker2ST#testRestoreOffsetsInConsumerGroup

**Comment by michalxo** [comment]
@strimzi-ci run tests test_only profile=regression testcase=MirrorMaker2ST#testRestoreOffsetsInConsumerGroup

**Review by scholzj — APPROVED** [review]

File: systemtest/src/test/java/io/strimzi/systemtest/mirrormaker/MirrorMaker2ST.java [hunk 8]
```diff
@@ -893,6 +900,170 @@ void testConfigureDeploymentStrategy() {
...
+            .endSpec().build());
+
+        // MM2 Active (S) <-> Active (T) // direction S <- T mirroring
+        KafkaMirrorMaker2Resource.createAndWaitForReadiness(KafkaMirrorMaker2Resource.kafkaMirrorMaker2(mm2SrcTrgName, kafkaClusterSourceName, kafkaClusterTargetName, 1, false)
+                .editSpec()
+                .editFirstMirror()
+                    .editSourceConnector()
+                    .addToConfig("refresh.topics.interval.seconds", "1")
+                    .addToConfig("replication.factor", "1")
+                    .addToConfig("offset-syncs.topic.replication.factor", "1")
```
Comment (Frawless) [inline systemtest/src/test/java/io/strimzi/systemtest/mirrormaker/MirrorMaker2ST.java:960]: indent

File: systemtest/src/test/java/io/strimzi/systemtest/mirrormaker/MirrorMaker2ST.java [hunk 9]
```diff
@@ -893,6 +900,170 @@ void testConfigureDeploymentStrategy() {
...
+                    .endSourceConnector()
+                    .editCheckpointConnector()
+                    .addToConfig("refresh.groups.interval.seconds", "1")
+                    .addToConfig("sync.group.offsets.enabled", "true")
+                    .addToConfig("sync.group.offsets.interval.seconds", syncGroupOffsetsIntervalSeconds)
```
Comment (Frawless) [inline systemtest/src/test/java/io/strimzi/systemtest/mirrormaker/MirrorMaker2ST.java:965]: indent

File: systemtest/src/test/java/io/strimzi/systemtest/mirrormaker/MirrorMaker2ST.java [hunk 10]
```diff
@@ -893,6 +900,170 @@ void testConfigureDeploymentStrategy() {
...
+        LOGGER.info("Receive 40 msgs from mirrored topic on Target cluster");
+        KafkaBasicExampleClients internalClientTargetJob = initialInternalClientTargetJob.toBuilder().withMessageCount(40).build();
+        internalClientTargetJob.createAndWaitForReadiness(internalClientTargetJob.consumerStrimzi().build());
+        ClientUtils.waitForClientSuccess(targetConsumerName, NAMESPACE, 40);
+        JobUtils.deleteJobWithWait(NAMESPACE, targetConsumerName);
+
+        LOGGER.info("There should be no more messages to read. Try to consume at least 1 message. " +
+                "This client job should fail on timeout.");
+        initialInternalClientTargetJob.createAndWaitForReadiness(initialInternalClientTargetJob.consumerStrimzi().build());
+        assertThrows(WaitException.class, () -> ClientUtils.waitForClientSuccess(targetConsumerName, NAMESPACE, 1));
```
Comment (Frawless) [inline systemtest/src/test/java/io/strimzi/systemtest/mirrormaker/MirrorMaker2ST.java:1049]: assert here is not needed, the test will fail even without assert in that case

File: systemtest/src/test/java/io/strimzi/systemtest/mirrormaker/MirrorMaker2ST.java [hunk 11]
```diff
@@ -893,6 +900,170 @@ void testConfigureDeploymentStrategy() {
...
+
+        LOGGER.info("As it's Active-Active MM2 mode, there should be no more messages to read from Source cluster" +
+                " topic. This client job should fail on timeout.");
+        initialInternalClientSourceJob.createAndWaitForReadiness(initialInternalClientSourceJob.consumerStrimzi().build());
+        assertThrows(WaitException.class, () -> ClientUtils.waitForClientSuccess(sourceConsumerName, NAMESPACE, 1));
```
Comment (Frawless) [inline systemtest/src/test/java/io/strimzi/systemtest/mirrormaker/MirrorMaker2ST.java:1054]: Same as above

File: systemtest/src/test/java/io/strimzi/systemtest/mirrormaker/MirrorMaker2ST.java [hunk 12]
```diff
@@ -893,6 +900,170 @@ void testConfigureDeploymentStrategy() {
...
+    void waitForMM2OffsetIntervalMirroring(String timeout) {
+        // wait sync.group.offsets.interval.seconds to sync in target cluster
+        LOGGER.info("Wait {} secs -> sync.group.offsets.interval.seconds, check msgs in target cluster", timeout);
+        try {
+            Thread.sleep(Duration.ofSeconds(Long.parseLong(timeout)).toMillis());
+        } catch (InterruptedException e) {
+            LOGGER.error("Error while waiting on syncGroupOffsetsIntervalSeconds");
+        }
+    }
+
```
Comment (Frawless) [inline systemtest/src/test/java/io/strimzi/systemtest/mirrormaker/MirrorMaker2ST.java:1066]: Why is this needed? Isn't better option to extend the timeout for clients? Even if it's just for one sec?

**Review by Frawless — CHANGES_REQUESTED** [review]

File: systemtest/src/test/java/io/strimzi/systemtest/mirrormaker/MirrorMaker2ST.java (hunk 10 above)
Comment (michalxo) [reply systemtest/src/test/java/io/strimzi/systemtest/mirrormaker/MirrorMaker2ST.java:1049]: Idea is, that I expect client to fail, so this is how to check it - discussed with @im-konge

File: systemtest/src/test/java/io/strimzi/systemtest/mirrormaker/MirrorMaker2ST.java (hunk 12 above)
Comment (michalxo) [reply systemtest/src/test/java/io/strimzi/systemtest/mirrormaker/MirrorMaker2ST.java:1066]: Imo, this gives test more readability about what's going on. Sure I can prolong client's timeout.

Comment (Frawless) [reply systemtest/src/test/java/io/strimzi/systemtest/mirrormaker/MirrorMaker2ST.java:1066]: what about comment add just comment instead of this weird sleep? Or wait for some event

File: systemtest/src/test/java/io/strimzi/systemtest/mirrormaker/MirrorMaker2ST.java (hunk 10 above)
Comment (Frawless) [reply systemtest/src/test/java/io/strimzi/systemtest/mirrormaker/MirrorMaker2ST.java:1049]: We should keep the tests consistent so we should add this to all clients calls, or use just clients calls as we do in other tests. This exception catch doesn't look good to me.

Comment (michalxo) [reply systemtest/src/test/java/io/strimzi/systemtest/mirrormaker/MirrorMaker2ST.java]: Added new method `ClientUtils.waitForClientTimeout()`

Comment (Frawless) [reply systemtest/src/test/java/io/strimzi/systemtest/mirrormaker/MirrorMaker2ST.java]: I am really sorry but I miss understand the test. I thought that you are testing it in a different way. Your first solution was right. Could you please revert your changes? I am sorry for my suggestions.

Comment (Frawless) [reply systemtest/src/test/java/io/strimzi/systemtest/mirrormaker/MirrorMaker2ST.java]: @im-konge  do you prefer `waitForClientTimeout()` in `ClientUtils` or use `assertThrow(() ->... )` ? I am fine with both solution, but in case we will use the first one, we should align the other tests with it

Comment (im-konge) [reply systemtest/src/test/java/io/strimzi/systemtest/mirrormaker/MirrorMaker2ST.java]: If you don't want to revert it, you can keep the method and I'll change it later.

Comment (michalxo) [reply systemtest/src/test/java/io/strimzi/systemtest/mirrormaker/MirrorMaker2ST.java]: So I believe we all agree on using method `ClientUtils.waitForClientTimeout()` (@see-quick also).
That means we can merge this PR once all tests pass. :-)

**Review by Frawless — APPROVED** [review]

**Review by im-konge — APPROVED** [review]
